sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -235,6 +235,7 @@ object ScalaReflection extends ScalaReflection {
       case t if isSubtype(t, localTypeOf[java.lang.String]) =>
         createDeserializerForString(path, returnNullable = false)
 
+      case t if isSubtype(t, localTypeOf[CalendarInterval]) => path
```

Comment:
just for curiosity: why interval inside array/map/struct works before this PR?

Reply:
Do you mean we work for scala Products? `Seq(Array(something)).toDF` seems need that `something` to be java primitives or `scala.Product`s. Not yet support `CalendarInterval` or e.g....
```scala
scala> case class A(a: Int)
defined class A

scala> Seq(Array(A(1))).toDF
res8: org.apache.spark.sql.DataFrame = [value: array<struct<a:int>>]

scala> class B(a: Int)
defined class B

scala> Seq(Array(new B(1))).toDF
<console>:29: error: value toDF is not a member of Seq[Array[B]]
       Seq(Array(new B(1))).toDF
                            ^

scala> Seq(new java.math.BigDecimal(1)).toDF
res10: org.apache.spark.sql.DataFrame = [value: decimal(38,18)]

scala> Seq(Array(new java.math.BigDecimal(1))).toDF
<console>:28: error: value toDF is not a member of Seq[Array[java.math.BigDecimal]]
       Seq(Array(new java.math.BigDecimal(1))).toDF
```

Reply:
They don't work
```scala
scala> Seq(Map(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3) -> new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
java.lang.UnsupportedOperationException: Schema for type org.apache.spark.unsafe.types.CalendarInterval is not supported
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:735)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:689)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.deserializerForType(ScalaReflection.scala:164)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:55)
  at org.apache.spark.sql.SQLImplicits.newMapEncoder(SQLImplicits.scala:175)
  ... 49 elided

scala> Seq(Seq(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3)).toDF
     | ;
<console>:2: error: ')' expected but ';' found.
       ;
       ^

scala> Seq(Seq(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
java.lang.UnsupportedOperationException: Schema for type org.apache.spark.unsafe.types.CalendarInterval is not supported
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$schemaFor$1(ScalaReflection.scala:735)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:663)
  at org.apache.spark.sql.catalyst.ScalaReflection$.toCatalystArray$1(ScalaReflection.scala:430)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerFor$1(ScalaReflection.scala:448)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.serializerFor(ScalaReflection.scala:410)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerForType$1(ScalaReflection.scala:399)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:877)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:876)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:47)
  at org.apache.spark.sql.catalyst.ScalaReflection$.serializerForType(ScalaReflection.scala:391)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:54)
  at org.apache.spark.sql.SQLImplicits.newSequenceEncoder(SQLImplicits.scala:171)
  ... 49 elided

scala> Seq(Array(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
<console>:28: error: value toDF is not a member of Seq[Array[org.apache.spark.unsafe.types.CalendarInterval]]
       Seq(Array(new org.apache.spark.unsafe.types.CalendarInterval(1,2,3))).toDF
```

Reply:
Sorry, I mean, struct:
> CalenderInterval is available to be converted to/from internal Spark SQL representation when it is a member of a Scala's product type e.g tuples/ case class etc but not as a primitive type

Reply:
with PR https://github.com/apache/spark/pull/28198/files  and the existing logic in `serializerFor`
1. https://github.com/yaooqinn/spark/blob/6cd0bef7fe872b5d9472a2140db188fb9f1eede8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L436
2. https://github.com/yaooqinn/spark/blob/6cd0bef7fe872b5d9472a2140db188fb9f1eede8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L554